### DISCUSSION TITLE
Publishing M³ in app stores

### DISCUSSION THREAD

**Discussion by ux-git** [sircharlo/meeting-media-manager#3259] [Ideas] [answered]
Are there currently any circumstances or reasons that might lead you to consider publishing the app on the Microsoft App Store and Apple App Store (_yes, there's the paid developer account, which we all don't like)_? I recall you, @sircharlo, mentioning that it wasn’t in the plans at the time, but that was quite a while ago. Since then, the macOS installation process has become a bit more challenging for regular users, with all the terminal commands and such.

What’s your perspective on the current situation and potential future scenarios?

**Comment by sircharlo** [comment, marked as answer]
Reading this discourages me: [How to Release an Electron App on the Mac App Store
](https://gist.github.com/steve981cr/def310670dfd9ed1439bf31cc734f941)

The time, equipment and resources required to do this are all out of reach for the moment, unfortunately.

**Reply by ux-git** [reply]
Do you mean that it involves re-building the app, or what exactly is the most challenging?

**Reply by mtdvlpr** [reply]
The re-building itself is not challenging, but the whole process involves having an Apple developer account of ~$100/year, maintaining three signing certificates which expire every year, maintaining provisioning profiles which expire every year and being reviewed by Apple before a release becomes available.

This means annual costs, extra recurring maintenance and a significantly slower release cycle.

**Reply by ux-git** [reply]
Got it, thanks for the answer

**Reply by mtdvlpr** [reply]
This, combined with the fact that all A/V instructions are written for Windows, makes it too much effort for too little gain (in my opinion).

**Reply by sbscomp** [reply]
@mtdvlpr If someone (like me) were to offer to do this with their mac as a developer, would it be worth considering? I realize it adds some difficulty to the process, but the up-side for our friends who are using mac devices for one reason or another is tremendous.

**Comment by mtdvlpr** [comment]
After doing some research, I think that the best way for us to implement this, would be to not publish the app to the App Store, but to sign and notarize the app from GitHub. This would resolve the existing security problems and require almost no extra efforts once it's setup.

### Steps to perform

#### 1. Code Sign the app from GitHub

Code Signing is currently done by the user with the `codesign --force --deep --sign - "/Applications/Meeting Media Manager.app"` command. If we want to code sign the app ourselves we need a Developer ID Installer and Developer ID Application certificate.

#### 2. Notarize the app from GitHub

Notarizing the app takes care of the second command that the user currently needs to execute: `sudo xattr -r -d com.apple.quarantine "/Applications/Meeting Media Manager.app"`. Notarizing the app means that Apple has verified the app and it won't be quarantined.

To notarize the app we need the following variables: `APPLE_ID`, `APPLE_APP_SPECIFIC_PASSWORD`, and `APPLE_TEAM_ID`.

### Implications

#### Users

- No longer need additional steps/commands for macOS users
- Support for auto updates (similar to Windows)

#### Developers

- Rely on the continuing support of the person to provide the Apple Developer Account resources

#### Certificates (https://developer.apple.com/support/certificates/)

- If your Developer ID Application Certificate expires, users can still download, install, and run versions of your Mac applications that were signed with this certificate. However, you’ll need a new certificate to sign updates and new applications. If your certificate is revoked, users will no longer be able to install applications that have been signed with this certificate.
- If your Developer ID Installer Certificate expires, users can still install packages that were signed with this certificate as long as the package includes a trusted timestamp. Previously installed apps will continue to run. However, new installations won’t be possible until you have re-signed your installer package with a valid Developer ID Installer certificate. If your certificate is revoked, users will no longer be able to install applications that have been signed with this certificate.

### Resources

- https://www.electronforge.io/guides/code-signing/code-signing-macos#obtaining-signing-certificates
- https://medium.com/@sdqumar09/electron-macos-code-signing-github-actions-and-locally-e7f0c62211d6
- https://dev.to/rwwagner90/signing-electron-apps-with-github-actions-4cof
- https://github.com/electron/notarize?tab=readme-ov-file#usage-with-app-specific-password

@sbscomp, if you or someone else is willing to provide us with the necessary certificates and credentials, we would be able to implement this without changing the release procedure.

**Reply by mtdvlpr** [reply]
@sbscomp, @sircharlo, I'd love your feedback on this :)

**Reply by sbscomp** [reply]
I'm completely willing to help out, but would it be better to have one of your names on it rather than an account that I owned?

**Reply by sircharlo** [reply]
I have nothing against using a certificate generously provided by @sbscomp or someone else. 
I would just need to delve into how to set this up practically.

**Reply by mtdvlpr** [reply]
> I'm completely willing to help out, but would it be better to have one of your names on it rather than an account that I owned?

Sure, but I assume that it would require a dedicated developer account for M3? Of course that would be ideal, but someone would have to be willing to set that up and maintain it. If we could use an existing account, the owner would not have extra costs and would therefore be able to support us for longer.

What matters to us, is that we receive the necessary resources and that it's maintainable for the person long-term. We don't want to implement this if we have to disable it again within the year. That would be confusing for the non-technical macOS users who are now used to the Terminal commands.

**Comment by ux-git** [comment]
Have you guys considered something like this solution? https://signpath.org/

**Reply by mtdvlpr** [reply]
If I read their documentation correctly, we'd still need a paid developer account...

See https://about.signpath.io/blog/2024/11/29/building-trusted-software-for-macos

**Comment by sbscomp** [comment]
@mtdvlpr and @sircharlo I have created a certificate that can be used to sign the application.

**Reply by mtdvlpr** [reply]
@sbscomp, awesome! I have a couple of questions:

1. Did you create a "Developer ID Application" certificate?
2. Could you export it to a `.p12` file and send it to us, along with a password for the certificate?
3. Are you also willing to provide us with the credentials to notarize the app (APPLE_ID, APPLE_TEAM_ID, APPLE_APP_SPECIFIC_PASSWORD)?

See https://medium.com/@sdqumar09/electron-macos-code-signing-github-actions-and-locally-e7f0c62211d6 for more information.

**Reply by sbscomp** [reply]
I did create a Developer ID Application certificate. I can export and willing to provide the credentials to you. I'll work on getting that done today.

**Reply by sbscomp** [reply]
I've sent you an encrypted email with the certificate, I'll send you another one with the credentials as soon as I have them set up.

**Reply by mtdvlpr** [reply]
Thank you very much! I'm excited to implement this and provide a significantly better user experience for macOS users.

**Reply by sbscomp** [reply]
Encrypted email with credentials has been sent. :)

**Reply by mtdvlpr** [reply]
@sircharlo, could you configure the following secrets in the repo?

- MAC_CERTIFICATE_BASE64 (as shared)
- MAC_CERTIFICATE_PASSWORD (as shared)
- KEYCHAIN_PASSWORD (random string)
- APPLE_ID (as shared)
- APPLE_TEAM_ID (as shared)
- APPLE_APP_SPECIFIC_PASSWORD (as shared)

**Reply by mtdvlpr** [reply]
Continued in #3804

**Reply by sircharlo** [reply]
@sbscomp Thank you so much for your certificate! It's truly appreciated. 

I think something might have been updated in the Apple Developer agreement recently; we started getting these errors during the build process:

```
Error: Failed to notarize via notarytool.  Failed with unexpected result:
Error: HTTP status code: 403. A required agreement is missing or has expired. 
This request requires an in-effect agreement that has not been signed or has expired. 
Ensure your team has signed the necessary legal agreements and that they are not expired.
```

Is there anything we need to do on our end for this to be resolved?

**Comment by sbscomp** [comment]
I took care of the agreement approval; you should be good to go again.

Thanks!
-Doug

From: Charles Olivier Savignac ***@***.***>
Date: Tuesday, June 24, 2025 at 8:06 AM
To: sircharlo/meeting-media-manager ***@***.***>
Cc: Douglas Phillips ***@***.***>, Mention ***@***.***>
Subject: Re: [sircharlo/meeting-media-manager] Publishing M³ in app stores (Discussion #3259)

@sbscomp<https://github.com/sbscomp> Thank you so much for your certificate! It's truly appreciated.

I think something might have been updated in the Apple Developer agreement recently; we started getting these errors during the build process:

Error: Failed to notarize via notarytool.  Failed with unexpected result:

Error: HTTP status code: 403. A required agreement is missing or has expired.

This request requires an in-effect agreement that has not been signed or has expired.

Ensure your team has signed the necessary legal agreements and that they are not expired.

Is there anything we need to do on our end for this to be resolved?

**Reply by sircharlo** [reply]
Thank you so much @sbscomp !

**Reply by sircharlo** [reply]
Hi there @sbscomp 
It's me again 😄 
This started showing up again:

`Error: HTTP status code: 403. A required agreement is missing or has expired. This request requires an in-effect agreement that has not been signed or has expired. Ensure your team has signed the necessary legal agreements and that they are not expired.
`

I'm sorry to bug you with this...

**Comment by sbscomp** [comment]
I'll get it taken care of today.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by sbscomp** [comment]
This has been completed.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by sircharlo** [reply]
Thank you so much!

**Comment by sircharlo** [comment]
Hi there @sbscomp,

It feels like this comes up every 6 months...
Is it still feasible for you to sign the agreement?

If not, we'll look into our other options.
Thanks!

**Comment by sbscomp** [comment]
Done.

Thanks
-Doug

From: Charles Olivier Savignac ***@***.***>
Date: Tuesday, March 17, 2026 at 12:23
To: sircharlo/meeting-media-manager ***@***.***>
Cc: Douglas Phillips ***@***.***>, Mention ***@***.***>
Subject: Re: [sircharlo/meeting-media-manager] Publishing M³ in app stores (Discussion #3259)


Hi there @sbscomp<https://github.com/sbscomp>,

It feels like this comes up every 6 months...
Is it still feasible for you to sign the agreement?

If not, we'll look into our other options.
Thanks!

**Reply by sircharlo** [reply]
> Done. Thanks -Doug 

Thank you very much!

**Comment by sircharlo** [comment]
Hi @sbscomp ! Sorry to bother you, but it happened again:

    Error: HTTP status code: 403. A required agreement is missing or has expired. This request requires an in-effect agreement that has not been signed or has expired. Ensure your team has signed the necessary legal agreements and that they are not expired.

**Comment by sbscomp** [comment]
I've taken care of it.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by sircharlo** [reply]
Much thanks!
